### PR TITLE
Update snowflake system tags with hierarchy

### PR DESCRIPTION
**Pull request by usefulalgorithm**
### 🤔 Why?

[SC21047](https://app.shortcut.com/metaphor-data/story/21047/)

### 🤓 What?

In `_fetch_tags`, when a database-level or schema-level tag is found, it will be included in the MCE as a part of a `Hierarchy`, which has a `HierarchyLogicalID` that starts with `DataPlatform.SNOWFLAKE.value`.

I will modify this PR once #643 is merged. 

Need to create story for frontend work I think.

### 🧪 Tested?

Wrote unit test for this. Tested against local dev environment.

Also tested against modified snowflake env, added the following system tags:
截圖 2023-10-26 晚上11 33 51
After running the crawler, we will have the following hierarchies in the output json:
[res.json](https://github.com/MetaphorData/connectors/files/13179681/res.json)

### CODE REVIEW

File: metaphor/snowflake/auth.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ class SnowflakeKeyPairAuthConfig:
     # provide decryption passphrase if private key is encrypted
     passphrase: Optional[str] = None
 
-    @root_validator
+    @root_validator(pre=False, skip_on_failure=True)
```
Comment (usefulalgorithm) [inline metaphor/snowflake/auth.py:37]: no

File: metaphor/snowflake/auth.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ class SnowflakeAuthConfig(BaseConfig):
     # the query tags for each snowflake query the connector issues
     query_tag: Optional[str] = METAPHOR_DATA
 
-    @root_validator
+    @root_validator(pre=False, skip_on_failure=True)
```
Comment (usefulalgorithm) [inline metaphor/snowflake/auth.py:66]: no

File: metaphor/snowflake/extractor.py [hunk 3]
```diff
@@ -599,7 +616,7 @@ def build_table_url(account: str, full_name: str) -> str:
         )
 
     def _parse_accessed_objects(self, raw_objects: str) -> List[QueriedDataset]:
-        objects = parse_raw_as(List[AccessedObject], raw_objects)
+        objects = parse_raw_as(List[AccessedObject], raw_objects)  # type: ignore
```
Comment (usefulalgorithm) [inline metaphor/snowflake/extractor.py:619]: no

File: metaphor/snowflake/auth.py (hunk 1 above)
Comment (perepelitca) [reply metaphor/snowflake/auth.py:37]: ? 🙂

Comment (usefulalgorithm) [reply metaphor/snowflake/auth.py:37]: ah sorry, I put this here just to remind myself to get rid of this once pydantic v2 PR is merged - in v2 `root_validator` is deprecated, instead it's going to be `model_validator`

Comment (mars-lan) [reply metaphor/snowflake/auth.py:37]: Another approach is to base your PR directly on https://github.com/MetaphorData/connectors/tree/tsung-julii/sc-19184/update-to-pydantic-2-x

**Comment by elic-eon** [comment]
There are another PR changes, can you separate them?

**Comment by usefulalgorithm** [comment]
@elic-eon  @mars-lan  @perepelitca  this is ready for review~

File: metaphor/snowflake/extractor.py [hunk 4]
```diff
@@ -135,6 +140,7 @@ async def extract(self) -> Collection[ENTITY_TYPES]:
         entities: List[ENTITY_TYPES] = []
         entities.extend(datasets)
         entities.extend(chunk_query_logs(self._logs))
+        entities.extend(list(self._hierarchies.values()))
```
Comment (elic-eon) [inline metaphor/snowflake/extractor.py:143]: Update tests to cover new entities

File: metaphor/snowflake/extractor.py [hunk 5]
```diff
@@ -359,45 +365,92 @@ def _fetch_primary_keys(self, cursor: SnowflakeCursor) -> None:
...
             SELECT TAG_NAME, TAG_VALUE, DOMAIN, OBJECT_DATABASE, OBJECT_SCHEMA, OBJECT_NAME, COLUMN_NAME
             FROM snowflake.account_usage.tag_references
-            WHERE DOMAIN in ('TABLE', 'COLUMN')
+            WHERE DOMAIN in ('TABLE', 'COLUMN', 'DATABASE', 'SCHEMA')
             ORDER BY OBJECT_DATABASE, OBJECT_SCHEMA, OBJECT_NAME
             """
         )
 
-        for key, value, object_type, database, schema, table_name, column in cursor:
+        for key, value, object_type, database, schema, object_name, column in cursor:
```
Comment (elic-eon) [inline metaphor/snowflake/extractor.py:420]: What `object_name` would be for schema/database level tag?

File: metaphor/snowflake/extractor.py [hunk 6]
```diff
@@ -359,45 +365,92 @@ def _fetch_primary_keys(self, cursor: SnowflakeCursor) -> None:
...
+        if not object_name:
+            logger.error("Cannot extract tag without object name")
+            return
+
+        # Set hierarchy key and logical_id
+        path = [DataPlatform.SNOWFLAKE.value]
+        if object_type == "DATABASE":
+            path.extend([object_name])
+            hierarchy_key = dataset_normalized_name(object_name)
+        elif object_name is not None:
```
Comment (elic-eon) [inline metaphor/snowflake/extractor.py:391]: `object_name` should be truthy here

File: metaphor/snowflake/extractor.py [hunk 7]
```diff
@@ -359,45 +365,92 @@ def _fetch_primary_keys(self, cursor: SnowflakeCursor) -> None:
                 sql_schema.primary_key = []
             sql_schema.primary_key.append(column)
 
+    def _add_system_tag(
+        self,
+        database: Optional[str],
+        object_name: Optional[str],
+        object_type: str,
```
Comment (elic-eon) [inline metaphor/snowflake/extractor.py:372]: Maybe we can narrow down `object_type` to `database|schema`?

File: metaphor/snowflake/extractor.py [hunk 8]
```diff
@@ -359,45 +365,92 @@ def _fetch_primary_keys(self, cursor: SnowflakeCursor) -> None:
...
+            )
+
+        # `system_tags` is never None here
+        self._hierarchies[hierarchy_key].system_tags.append(
+            SystemTag(
+                key=tag_key,
+                system_tag_source=SystemTagSource.SNOWFLAKE,
+                value=tag_value,
+            )
+        )
```
Comment (elic-eon) [inline metaphor/snowflake/extractor.py:408]: Maybe you can use [dict.setdefault](https://docs.python.org/3/library/stdtypes.html?highlight=setdefault#dict.setdefault) to set the default value and get the reference to append the system tag.

File: metaphor/snowflake/extractor.py (hunk 4 above)
Comment (usefulalgorithm) [reply metaphor/snowflake/extractor.py:143]: I don't think the existing tests ever run `extractor.extract`, so none of these entities are actually covered. I'm currently having trouble on this...

File: metaphor/snowflake/extractor.py (hunk 5 above)
Comment (usefulalgorithm) [reply metaphor/snowflake/extractor.py:420]: for schema level tags it's schema name, and vice versa

File: metaphor/snowflake/extractor.py (hunk 4 above)
Comment (elic-eon) [reply metaphor/snowflake/extractor.py:143]: NVM, the test you added should be good enough

File: poetry.lock
Comment (elic-eon) [inline poetry.lock:1]: Normally, I use `poetry lock --no-update` and `poetry install -E all` to update the package. It won't update all dependencies every time.

Comment (usefulalgorithm) [reply poetry.lock:1]: got it

File: metaphor/snowflake/extractor.py (hunk 4 above)
Comment (elic-eon) [inline metaphor/snowflake/extractor.py:143]: I think `extend` can take `iterable`?

**Review by elic-eon — APPROVED** [review]
LGTM!

**Pull request merged**